lib/strings.nix

```
@@ -355,6 +355,21 @@ rec {
   */
   escapeC = list: replaceStrings list (map (c: "\\x${ toLower (lib.toHexString (charToInt c))}") list);
 
+  /* Escape the string so it can be safely placed inside a URL
+     query.
+
+     Type: escapeQuery = string -> string
+
+     Example:
+       escapeQuery "foo bar"
+       => "foo%20bar"
+  */
+  escapeQuery = s:
+    let
+      list = [ "!" "#" "$" "&" "'" "(" ")" "*" "+" "," "/" ";" "=" "?" "@" "[" "]" " " "%" "." "<" ">" "\\" "^" "_" "`" "{" "|" "}" "~" ];
+    in
+      replaceStrings list (map (c: "%${lib.toHexString (charToInt c)}") list) s;
```

```suggestion
  escapeURL = let
    unreserved = [ "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z" "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s" "t" "u" "v" "w" "x" "y" "z" "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" "-" "_" "." "~" ];
    escaped = builtins.removeAttrs asciiTable unreserved;
  in replaceStrings (builtins.attrNames escaped) (lib.mapAttrsToList (_: c: "%${lib.toHexString c}") escaped);
```

(where `asciiTable` is the binding from `charToInt`, moved to the top of the file)

Escape all known ASCII characters that *aren't* in the unreserved set. This is less error-prone than listing the disallowed characters. Note that `~` is unreserved.

Note that in theory non-ASCII characters should be encoded as well (all bytes not in the unreserved set, really), but Nix makes this hard so this is fine.

About the name: I suggest `escapeURL` since this is not specific to query strings, it can be used to encode any string meant to be included in a URL.

---

What is the benefit of declaring `table` at the top of the file? Is it to do with caching?

---

Currently the file would be imported every time the function is called. You can avoid that by η-reducing the `s` argument away, but it would still be imported twice if one uses both `charToInt` and `escapeURL`. If you bind it at the top of the file, it should only be imported once.

---

Ok I guess Nix does file caching anyway, so the remaining argument is to avoid code duplication.